File: benchmarks/python/08_codelist.py
"""Benchmark 08: Codelist Generation — omopy.codelist.get_candidate_codes()"""

import polars as pl
from helpers import Timer, connect_cdm, save_result, save_timing

from omopy.codelist import get_candidate_codes

print("=== 08: Codelist ===")
t = Timer()
cdm = connect_cdm()

codes = get_candidate_codes(
    cdm, keywords=["coronary"], domains=["Condition"], include_descendants=True
)
# codes is a Codelist (dict-like) — flatten to a DataFrame
rows = []
for name, concept_ids in codes.items():
    for cid in concept_ids:
        rows.append({"codelist_name": name, "concept_id": cid})
df = pl.DataFrame(rows)

save_result(df, "08_codelist")
save_timing("08_codelist", t.elapsed())
print(f"Done in {t.elapsed():.2f} seconds")

File: benchmarks/r/08_codelist.R
# Benchmark 08: Codelist Generation
# R equivalent: CodelistGenerator::getCandidateCodes()

source("benchmarks/r/00_helpers.R")
library(CodelistGenerator)
cat("=== 08: Codelist ===\n")
t0 <- proc.time()

con <- dbConnect(duckdb(), dbdir = DB_PATH)
cdm <- cdmFromCon(con = con, cdmSchema = "main", writeSchema = "main", cdmName = "synthea_1k")

codes <- getCandidateCodes(
  cdm = cdm,
  keywords = "coronary",
  domains = "Condition",
  includeDescendants = TRUE
)

save_result(as.data.frame(codes), "08_codelist")
elapsed <- (proc.time() - t0)[["elapsed"]]
save_timing("08_codelist", elapsed)
cdmDisconnect(cdm)
cat("Done in", elapsed, "seconds\n")